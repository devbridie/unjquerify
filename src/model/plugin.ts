import {JQueryExpressionType} from "./matchers/jquery-expression-type";
import {ReturnType} from "./return-types/return-type";
import {Expression, SpreadElement, Statement} from "babel-types";
import {Scope} from "babel-traverse";

export interface Plugin {
    matchesExpressionType: JQueryExpressionType;
    applicableWithArguments: (args: Array<Expression | SpreadElement>) => boolean;
    returnType: ReturnType;
    replaceWith: (element: Expression, args: Expression[], scope: Scope) => Expression | Statement[];
    escapeFromChain?: true;
}
