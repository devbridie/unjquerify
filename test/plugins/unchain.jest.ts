import "jest";
import * as babel from "babel-core";
import {jQueryExpressionPlugin} from "../../src/plugins/jquery-expression.plugin";
import {AddClassPlugin} from "../../src/plugins/attributes/addclass.plugin";
import {QuerySelectorAllPlugin} from "../../src/plugins/selectors/querySelectorAll.plugin";
import {FindPlugin} from "../../src/plugins/traversing/tree-traversal/find.plugin";
import {CssGetPlugin} from "../../src/plugins/manipulation/style-properties/css.get.plugin";
import {HidePlugin} from "../../src/plugins/effects/basics/hide.plugin";
import {CssSetPlugin} from "../../src/plugins/manipulation/style-properties/css.set.plugin";

describe("unchain", () => {
    describe("as statement", () => {
        test("transforms mutationless expression correctly", () => {
            const example = `$("a").addClass("clazz").hide().css("font-size",12);`;
            const plugins = [AddClassPlugin, QuerySelectorAllPlugin, HidePlugin, CssSetPlugin];
            const transformed = babel.transform(example, {plugins: [jQueryExpressionPlugin(plugins)]}).code;
            expect(transformed).toMatchSnapshot();
        });

        test("transforms mutated expression correctly", () => {
            const example = `$("div").addClass("clazz").find("a").css("font-size",12);`;
            const plugins = [AddClassPlugin, QuerySelectorAllPlugin, FindPlugin, CssSetPlugin];
            const transformed = babel.transform(example, {plugins: [jQueryExpressionPlugin(plugins)]}).code;
            expect(transformed).toMatchSnapshot();
        });
    });

    describe("as assignment rval", () => {
        test("transforms mutationless expression correctly value", () => {
            const example = `const as = $("a").addClass("clazz").hide().css("font-size",12);`;
            const plugins = [AddClassPlugin, QuerySelectorAllPlugin, CssSetPlugin, HidePlugin];
            const transformed = babel.transform(example, {plugins: [jQueryExpressionPlugin(plugins)]}).code;
            expect(transformed).toMatchSnapshot();
        });

        test("transforms mutationless expression correctly", () => {
            const example = `const fontSize = $("a").addClass("clazz").hide().css("font-size");`;
            const plugins = [AddClassPlugin, QuerySelectorAllPlugin, CssGetPlugin, HidePlugin];
            const transformed = babel.transform(example, {plugins: [jQueryExpressionPlugin(plugins)]}).code;
            expect(transformed).toMatchSnapshot();
        });

        test("transforms mutated expression correctly", () => {
            const example = `const fontSize = $("div").addClass("clazz").css("font-size");`;
            const plugins = [AddClassPlugin, QuerySelectorAllPlugin, CssGetPlugin];
            const transformed = babel.transform(example, {plugins: [jQueryExpressionPlugin(plugins)]}).code;
            expect(transformed).toMatchSnapshot();
        });

        test("transforms child mutated expression correctly as value", () => {
            const example = `const fontSize = $("div").addClass("clazz").find("a").css("font-size");`;
            const plugins = [AddClassPlugin, QuerySelectorAllPlugin, FindPlugin, CssGetPlugin];
            const transformed = babel.transform(example, {plugins: [jQueryExpressionPlugin(plugins)]}).code;
            expect(transformed).toMatchSnapshot();
        });
    });

    describe("short chain", () => {
        test("transforms self", () => {
            const example = `const as = $("a").addClass("clazz");`;
            const plugins = [AddClassPlugin, QuerySelectorAllPlugin];
            const transformed = babel.transform(example, {plugins: [jQueryExpressionPlugin(plugins)]}).code;
            expect(transformed).toMatchSnapshot();
        });
    });
});
