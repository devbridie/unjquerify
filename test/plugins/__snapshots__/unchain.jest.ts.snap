// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`unchain as assignment rval transforms child mutated expression correctly as value 1`] = `
"const _start = document.querySelectorAll(\\"div\\");

_start.forEach(_element => _element.classList.add(\\"clazz\\"));

const _chain = [].concat(..._start.map(_element2 => _element2.querySelectorAll(\\"a\\")));

const fontSize = getComputedStyle(_chain[0])[\\"font-size\\"];"
`;

exports[`unchain as assignment rval transforms mutated expression correctly 1`] = `
"const _start = document.querySelectorAll(\\"div\\");

_start.forEach(_element => _element.classList.add(\\"clazz\\"));

const fontSize = getComputedStyle(_start[0])[\\"font-size\\"];"
`;

exports[`unchain as assignment rval transforms mutationless expression correctly 1`] = `
"const _start = document.querySelectorAll(\\"a\\");

_start.forEach(_element => _element.classList.add(\\"clazz\\"));

_start.forEach(_element2 => _element2.style.display = \\"none\\");

const fontSize = getComputedStyle(_start[0])[\\"font-size\\"];"
`;

exports[`unchain as assignment rval transforms mutationless expression correctly value 1`] = `
"const as = document.querySelectorAll(\\"a\\");

_start.forEach(_element => _element.classList.add(\\"clazz\\"));

_start.forEach(_element2 => _element2.style.display = \\"none\\");

_start.forEach(_element3 => _element3.setAttribute(\\"font-size\\", 12));"
`;

exports[`unchain as statement transforms mutated expression correctly 1`] = `
"const _start = document.querySelectorAll(\\"div\\");

_start.forEach(_element2 => _element2.classList.add(\\"clazz\\"));

const _chain = [].concat(..._start.map(_element3 => _element3.querySelectorAll(\\"a\\")));

_chain.forEach(_element4 => _element4.setAttribute(\\"font-size\\", 12));"
`;

exports[`unchain as statement transforms mutationless expression correctly 1`] = `
"const _start = document.querySelectorAll(\\"a\\");

_start.forEach(_element2 => _element2.classList.add(\\"clazz\\"));

_start.forEach(_element3 => _element3.style.display = \\"none\\");

_start.forEach(_element4 => _element4.setAttribute(\\"font-size\\", 12));"
`;

exports[`unchain multi statement chain transforms properly 1`] = `
"const divs = document.querySelectorAll(\\"div\\");
divs.forEach(_element => _element.classList.add(\\"visited\\"));
const links = [].concat(...divs.map(_element2 => _element2.querySelectorAll(\\"a\\")));

links.forEach(_element3 => _element3.textContent = \\"A!\\");"
`;

exports[`unchain short chain transforms self 1`] = `
"const as = document.querySelectorAll(\\"a\\");
as.forEach(_element => _element.classList.add(\\"clazz\\"));"
`;
